base/shared/src/main/scala/scalaz/core/eq.scala

```diff
@@ -1,14 +1,16 @@
 package scalaz
 package core
 
+import scala.AnyRef
 import scala.language.experimental.macros
 
 trait EqClass[A] {
   def equal(first: A, second: A): Boolean
```

Comment:
I think you can get the exact same benefit with

```scala
trait EqClass[A] {
  final def equal(a1: A, a2: A): Boolean = 
    (a1.asInstanceOf[AnyRef] eq a2.asInstanceOf[AnyRef]) || equalImpl(a1, a2)
  protected[this] def equalImpl(a1: A, a2: A): Boolean
}
```

As @edmundnoble pointed out, `AnyRef` is (mostly) a lie; inside a method, unbounded generic type parameters will all be (at runtime) `Object`. (The "mostly" occurs when you write `def wat[A](arr: Array[A]) = ...`; that's different than `def why[A <: AnyRef](arr: Array[A])` since `arr` in `why` will be `Object[]`, while `arr` in `wat` will be `Object` (lub of `int[]` with `Object[]` with `long[]` with ...)).

Reply:
This was my original suggestion. Now I question whether it will work with `@specialized` annotation; rather, I suspect it will work but involve boxing.